Refactor some duplicated methods in *action classes #6448

AnaghHegde opened the pull request:
Fixes #6448

<!--
Please follow the naming conventions in the "guidelines for contributing" link above or the pull request may be rejected.

Also, do consider outlining the solution taken as doing so will likely help in the reviewing process
(e.g., when the pull request involves non-trivial changes)
-->

samsontmr commented on src/main/java/teammates/ui/controller/Action.java:
```diff
@@ -577,6 +585,126 @@ private boolean isMasqueradeModeRequested(AccountAttributes loggedInUser, String
     }
     
+    /**
+     *Common method to Get the feedback data  
```
Your build is failing because of trailing spaces in this line

AnaghHegde replied:
@samsontmr Its a comment, does it really matter?

samsontmr replied:
Well I guess it mattered to Travis :P

AnaghHegde commented:
@wkurniawan07 @samsontmr Please delay the review.

AnaghHegde commented:
@wkurniawan07 @Neurrone  Is the approach is right ?. Currently i am working on failing test cases.

Neurrone commented on src/main/java/teammates/ui/controller/Action.java:
```diff
@@ -576,10 +584,131 @@ private boolean isMasqueradeModeRequested(AccountAttributes loggedInUser, String
+        // getting the class name from which it's being called
+        String className = new Exception().getStackTrace()[1].getClassName();
+
+        if ("InstructorFeedbackEditSaveAction".equals(className)) {
```
Like in your previous PR, avoid doing explicit checking of class names as that makes the implementation brittle. Use inheritance to add the subclass-specific logic.

AnaghHegde replied:
@Neurrone 
>Use inheritance to add the subclass-specific logic.

What exactly are you suggesting?. Can you be little more specific?

thenaesh replied:
@AnaghHegde 
Perhaps this code you're writing belongs in the `InstructorFeedbackEditSaveAction` class itself? Right now you're writing code that only applies to that particular subclass, but you're putting it in the `Action` class, which introduces an unnecessary coupling in the opposite direction (now `Action` needs to know about `InstructorFeedbackEditSaveAction`).

You can override this method (since you've made it `protected`, after all) in the `InstructorFeedbackEditSaveAction` class, after defining a more general method here in the `Action` class.

Also, in the _occasional_ case when you need to actually get the name of the current class, use `this.getClass()`, which returns an object of type `Class<?>` which you can compare directly, or use `this.getClass().getName()` to obtain a string representation. There's no need to construct an exception and get its stack trace; you're misusing the `Exception` object to do something it is not really intended to do, and therefore making your code harder to understand.

Hope this helps. :)

AnaghHegde replied:
@thenaesh 
>You can override this method (since you've made it protected, after all) in the InstructorFeedbackEditSaveAction class, after defining a more general method here in 

This one i thought of earlier this is not gonna owrk in my case because there difference in the  method definition like one class needs one variable other class doesn't need.So i didn't find any ways to write general definition.

But,
>this.getClass(), which returns an object of type Class<?> which you can compare directly, or use this.getClass().getName() to obtain a string representation. 

It will definately help me.Thanks a lot

thenaesh replied:
@AnaghHegde 
How about splitting `extractFeedbackSessionData` into several smaller methods? Then perhaps you will have one small method whose actions need to differ based on the class they are called on, which can be easily overriden.

In particular,
```
if ("InstructorFeedbackEditSaveAction".equals(className)) {
   newSession.setCreatorEmail(getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_CREATOR));
} else {
  newSession.setCreatedTime(new Date());
}
```
can be extracted into a smaller method, which does `newSession.setCreatedTime(new Date())` in the base class and can be overriden to do `newSession.setCreatorEmail(getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_CREATOR))` inside `InstructorFeedbackEditSaveAction`. This method can then be called from `extractFeedbackSessionData`, and binds to either the base class version or the subclass version depending on the type of the object from which it is being called.

Neurrone commented on src/main/java/teammates/ui/controller/InstructorFeedbackAddAction.java:
```diff
@@ -45,7 +38,8 @@ protected ActionResult execute() {
 
-        FeedbackSessionAttributes fs = extractFeedbackSessionData();
+        InstructorFeedbackAddAction newSession = new InstructorFeedbackAddAction();
+        FeedbackSessionAttributes fs = newSession.extractFeedbackSessionData();
```
You should be able to use the method directly since this class inherits from Action

AnaghHegde replied:
You need to have a subclass object to call superclass method. Else you need to override the superclass method and have to use `super` which gives error in the build which says **overriding method merely calls super** .

chowyb replied:
You do not need to create a new object for this, you can simply call the method directly. Other classes (eg `InstructorCourseInstructorAddAction`) do this just fine.

samsontmr replied:
I believe what you are doing now is creating a new `InstructorFeedbackAddAction` object, then calling `extractFeedbackSessionData()` on that new object while calling `.execute()` on a `InstructorFeedbackAddAction` object. @Neurrone is right, you should be calling `extractFeedbackSessionData()` directly instead.

AnaghHegde replied:
ok. Now i realized what @Neurrone was trying to say.

Neurrone commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditSaveAction.java:
```diff
@@ -36,7 +27,8 @@ protected ActionResult execute() throws EntityDoesNotExistException {
                 Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION);
         
         InstructorFeedbackEditPageData data = new InstructorFeedbackEditPageData(account);
-        FeedbackSessionAttributes feedbackSession = extractFeedbackSessionData();
```
Similarly here, you can use the method directly due to inheritance.

thenaesh commented on src/main/java/teammates/ui/controller/Action.java:
```diff
@@ -576,10 +584,131 @@ private boolean isMasqueradeModeRequested(AccountAttributes loggedInUser, String
+        // getting the class name from which it's being called
+        String className = this.getClass().getName();
+
+        if ("InstructorFeedbackEditSaveAction".equals(className)) {
```
@AnaghHegde 
How about splitting `extractFeedbackSessionData` into several smaller methods? Then perhaps you will have one small method whose actions need to differ based on the class they are called on, which can be easily overriden.

In particular,
```
if ("InstructorFeedbackEditSaveAction".equals(className)) {
   newSession.setCreatorEmail(getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_CREATOR));
} else {
  newSession.setCreatedTime(new Date());
}
```
can be extracted into a smaller method, which does `newSession.setCreatedTime(new Date())` in the base class and can be overriden to do `newSession.setCreatorEmail(getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_CREATOR))` inside `InstructorFeedbackEditSaveAction`. This method can then be called from `extractFeedbackSessionData`, and binds to either the base class version or the subclass version depending on the type of the object from which it is being called.

AnaghHegde replied:
@thenaesh 
>can be extracted into a smaller method, which does newSession.setCreatedTime(new Date()) in the base class

  These atrributes has to created one after other  means - before doing `newSession.setCreatorEmail()`  say i have to do `setCreateTime()` ?. If that doesn't affect the functionality of the method then i can put those attributes in the subclass itself.

>This method can then be called from extractFeedbackSessionData, and binds to either the base class version or the subclass version **depending on the type of the object from which it is being called.**

Again to check the type of the object called i have to use the classname right ?

thenaesh replied:
@AnaghHegde 
No, that is done automatically due to polymorphism.

If you have
```
class Action {
  protected void doSomething() {
    // do something general
  }
  protected FeedbackSessionAttributes extractFeedbackSessionData() {
    this.doSomething();
  }
}
```
and
```
class InstructorFeedbackEditSaveAction extends Action {
  @Override
  protected void doSomething() {
    // do something specific to this subclass
  }
}
```
, then the subclass version of `doSomething` will automatically be called when the `Action` object, whose `extractFeedbackSessionData` method is being called, is actually an `InstructorFeedbackEditSaveAction` object. No explicit checking of the object type needs to occur, so there is no need to get any classname.

Of course, this is a contrived example, but it should illustrate my point. Think about how the  code that is specific to the `InstructorFeedbackEditSaveAction` can be extracted in this manner.

AnaghHegde replied:
@thenaesh Thanks for the help. Finally `this.doSomething()` cleared my all confusions.

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackAddAction.java:
```diff
@@ -2,32 +2,25 @@
-import com.google.appengine.api.datastore.Text;
 import com.google.gson.reflect.TypeToken;
 
 public class InstructorFeedbackAddAction extends InstructorFeedbacksPageAction {
```
Not your fault, but this is a wrong inheritance as only `Action` and `*AbstractAction` can be extended. Can you fix this?

AnaghHegde commented:
@wkurniawan07 Sorry for replying late , some clarifications are needed just to clear my confusion.

1.
> make the two children classes with common method inheriting that abstract class.

You are talking about these two classes right? 1.InstructorFeedbackAddAction  2.InstructorFeedbackEditSaveAction

2.
>IntructorFeedbacksPageAction inheriting the new AbstractAction class as well and move the reused methods to that class.
 
What do you mean by `move the reused methods to that class`. 

3.
>Not your fault, but this is a wrong inheritance as only Action and *AbstractAction can be extended. Can you fix this?

before this you said `IntructorFeedbacksPageAction inheriting the new AbstractAction class as well and move the reused methods to that class.`

So what i am thinking is : remove the inheritance relation between 1.`IntructorFeedbacksPageAction`
and 2.`IntructorFeedbackPageAction` by adding what all the methods required by class 2 to class 1.

Then make it to inherit new `Abstraction class` and this new class inherits from `Action` class .

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,185 @@
+            newSession.setCreatorEmail(getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_CREATOR));
+        } else {
+            newSession.setCreatedTime(new Date());
+        }
```
Now we can make use of the `abstract`-ness of this class to get rid of this `if` block.
```java
protected abstract void doSomething(FeedbackSessionAttributes newSession);
...
doSomething(newSession);
...
```
and implement `doSomething` accordingly in all the implementing classes. Of course, rename `doSomething` to something else more meaningful.

Neurrone commented on src/main/java/teammates/ui/controller/Action.java:
```diff
@@ -576,9 +576,9 @@ private boolean isMasqueradeModeRequested(AccountAttributes loggedInUser, String
                && !loggedInUser.googleId.equals(requestedUserId);
     }
-    
+
```
Remove the newline to minimize unnecessary changes

Neurrone commented on src/main/java/teammates/ui/controller/Action.java:
```diff
@@ -576,9 +576,9 @@ private boolean isMasqueradeModeRequested(AccountAttributes loggedInUser, String
+
     // ===================== Utility methods used by some child classes========
-    
+
```
Here too.

AnaghHegde replied:
@Neurrone Earlier i did changes. So after that i reverted back to original file only.Please take a look at this image. Its same as master.
![temp](https://cloud.githubusercontent.com/assets/12005551/22465153/46556a58-e7e0-11e6-9e61-834583066aca.png)

wkurniawan07 replied:
Just a quick advice that "everything looks the same" based on human eyesight can be dangerous when computer files are involved. There are things such as trailing whitespaces, different line endings, different encodings, etc. For example, these two are *not* the same:
```
	Some text
```
```
        Some text
```

Neurrone replied:
The original line had 4 spaces, which got replaced with an empty line.

AnaghHegde commented:
@wkurniawan07 @Neurrone @samsontmr @thenaesh @chowyb 

Please delay the review some small changes are still needed.

AnaghHegde commented:
@wkurniawan07  can you please tell me why the build is failing. What should i have to do!?.

samsontmr commented:
@AnaghHegde is your branch updated?

AnaghHegde commented:
Yeah its even with the master

Neurrone commented:
Think its a Travis problem getting the dependencies. I just had a build fail as well.

wkurniawan07 commented:
@AnaghHegde earlier this morning Travis had some problems, which has been resolved now. Your build still fails so you need to push more fixes.

wkurniawan07 commented on src/test/java/teammates/test/cases/browsertests/InstructorFeedbackPageUiTest.java:
```diff
@@ -279,7 +279,7 @@ public void testAddAction() throws Exception {
         newSession.setGracePeriod(15);
         
-        newSession.setFeedbackSessionType(FeedbackSessionType.PRIVATE);
+        newSession.setFeedbackSessionType(FeedbackSessionType.STANDARD);
```
You should not need to change this or any HTML file for testing purpose. Refactoring should not modify the action's behaviour.

AnaghHegde commented:
@wkurniawan07  Ready for review. 
I have added 2 methods because some of the attributes which depends on (take look at line 113) `List<String>` which depends on `String[]` array. . 

I tried to put all of them in a single method but some of tests where failing and @wkurniawan07 said  **Refactoring should not modify the action's behavior**  i have added 2 methods.

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,168 @@
+
+    protected FeedbackSessionAttributes extractFeedbackSessionDataHelper(
+            FeedbackSessionAttributes newSession, List<String> sendRemainderEmailsList) {
+        return this.extractFeedbackSessionDataHelper(newSession, sendRemainderEmailsList);
```
For these two helper methods:
1. Please rename with a better name. `*Helper` does not help in telling what the methods do.
2. Make them `abstract` instead.
3. See if you can combine them into one method.

AnaghHegde replied:
>See if you can combine them into one method.

I tried to put all of them in a single method but some of tests where failing and @wkurniawan07  said Refactoring should not modify the action's behavior i have added 2 methods instead of one method.

wkurniawan07 replied:
"Some tests are failing" is not a good enough reason for things to not work. If you can back up with solid reasons, e.g. "we need two methods because they operate on different states of the session object: ... (explanations)" (just an example), then I will accept that you need two methods.

AnaghHegde replied:
@wkurniawan07 I believe followings are the reasons for having 2 methods : 

* It affects the code readability. Suppose if someone in the future wants to some changes then    dev won't understand why this(`List<String>` and `String[]` )are at the start because the `const` used in creating array and the array name `senderReminderEmail` which says more than enough.

  
* First i wasn't sure whether the order of attributes creation matters but now due to the failing tests and the build log it matters.  

* Since the method is `extractFeedbackSessionData` why do we have to send the reminder email about the feedback we need to send it to only those who haven't submitted the feedback. If we create the `List<String>` and `String[]` at starting this happens.

* If you wonder why can't we add it at the end there are some attributes like `setCreatedTime` which has to be created in the start.

So i think we can't handle all these and have a single method.

wkurniawan07 replied:
> It affects the code readability and changes will not be minimum.

In what way does having just one method affect code readability?
Having two functions with vague names (`*Helper`), overloaded but doing entirely different things (overloaded functions should do similar things), now that affects readability very negatively.

> i wasn't sure whether the order of attributes creation matters but now due to the failing tests and the build log(commit 581f184) it matters.

Quoting what I said before: failing tests is not a good reason. "It matters": in what ways do it matter?

> In the method extractFeedbackSessionData why do we have to send the reminder email about the feedback to all ?, we need to send it to only those who haven't submitted the feedback. If we create the List<String> and String[] at starting this happens.

I don't get what this means.

> there are some attributes like setCreatedTime which has to be created in the start. Remaining attributes which uses time depends on this.

Which lines give you this impression? I am pretty sure this is not the case, although I might be wrong.

Also, do address the first two items of my very first comment.

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,168 @@
+                sendReminderEmailsArray == null ? new ArrayList<String>()
+                        : Arrays.asList(sendReminderEmailsArray);
+        newSession = extractFeedbackSessionDataHelper(newSession, sendReminderEmailsList);
+        newSession = extractFeedbackSessionDataHelper(newSession, sendReminderEmailsList);
```
This method is being called twice?

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,168 @@
+
+    /*
+     * InstructorFeedbacksPageAction and InstructorFeedbackAddAction class
+     * related methods
```
Remove this. Methods should not be documented on where they are used.

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,168 @@
+     * @return
+     */
+    protected HashMap<String, InstructorAttributes> loadCourseInstructorMap(boolean omitArchived) {
+        HashMap<String, InstructorAttributes> courseInstructorMap = new HashMap<String, InstructorAttributes>();
```
I note that it is the fault existing from the old code, but let's do the following:
1. Either remove the `@return` tag on the Javadoc, or add a meaningful information in it.
1. protected ~~Hash~~Map&lt;String, InstructorAttributes&gt; loadCourseInstructorMap
1. ~~Hash~~Map&lt;String, InstructorAttributes&gt; courseInstructorMap = ...

If you want to know more about the last two, google the term "loose coupling".

AnaghHegde replied:
Loose coupling is preferred over tight coupling right?.

Neurrone replied:
Yes

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,168 @@
+public abstract class InstructorFeedbackAbstractAction extends Action {
+
+    /**
+     * Common method to Get the feedback data
```
Remove this.

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackAddAction.java:
```diff
@@ -147,88 +143,19 @@ private void createTemplateFeedbackQuestions(String courseId, String feedbackSes
 
-    private FeedbackSessionAttributes extractFeedbackSessionData() {
-        //TODO assert parameters are not null then update test
-        //TODO make this method stateless
```
These two TODOs are not done yet. Add them to the method in the abstract class.

AnaghHegde commented:
@wkurniawan07 Ready for review.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,160 @@
+                getRequestParamValues(Const.ParamsNames.FEEDBACK_SESSION_SENDREMINDEREMAIL);
+        List<String> sendReminderEmailsList =
+                sendReminderEmailsArray == null ? new ArrayList<String>()
+                        : Arrays.asList(sendReminderEmailsArray);
```
`:` should be under `?` from the line above

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,160 @@
+            break;
+        default:
+            log.severe("Invalid sessionVisibleFrom setting in creating "
+                    + newSession.getIdentificationString());
```
`+` should be under `"` of the previous line

AnaghHegde replied:
this is after i did format of the source code using eclipse.

samsontmr replied:
Eclipse doesn't always get it right.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,160 @@
+            break;
+        default:
+            log.severe("Invalid resultsVisibleFrom setting in creating"
+                    + newSession.getIdentificationString());
```
`+` should be under `"` of the previous line

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,160 @@
+        newSession.setClosingEmailEnabled(sendReminderEmailsList.contains(EmailType.FEEDBACK_CLOSING
+                .toString()));
+        newSession.setPublishedEmailEnabled(sendReminderEmailsList.contains(EmailType.FEEDBACK_PUBLISHED
+                .toString()));
```
Readability/formatting of these four lines can be improved

AnaghHegde replied:
you meant formatting right? or you have something else in mind?

samsontmr replied:
Yup

AnaghHegde replied:
ok i will change those

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,160 @@
+    protected Map<String, InstructorAttributes> loadCourseInstructorMap(boolean omitArchived) {
+        Map<String, InstructorAttributes> courseInstructorMap = new HashMap<String, InstructorAttributes>();
+        List<InstructorAttributes> instructors = logic.getInstructorsForGoogleId(account.googleId,
+                omitArchived);
```
`omitArchived);` should be aligned with `account.googleId`

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditSaveAction.java:
```diff
@@ -63,94 +60,21 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+    protected FeedbackSessionAttributes extractFeedbackSessionDataHelper(
+            FeedbackSessionAttributes newSession, List<String> sendReminderEmailsList) {
+        newSession.setOpeningEmailEnabled(sendReminderEmailsList.contains(EmailType.FEEDBACK_OPENING
+                .toString()));
```
Here is another one, the periods should be aligned

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,160 @@
+        newSession.setClosingEmailEnabled(sendReminderEmailsList.contains(EmailType.FEEDBACK_CLOSING
+                                                                                    .toString()));
+        newSession.setPublishedEmailEnabled(sendReminderEmailsList.contains(EmailType.FEEDBACK_PUBLISHED
+                                                                                      .toString()));
```
Periods should be aligned

damithc replied:
In fact, I suggest this (prefer to insert the line break at a higher level in the syntax tree):
```java
newSession.setPublishedEmailEnabled(
        sendReminderEmailsList.contains(EmailType.FEEDBACK_PUBLISHED.toString()));
```

AnaghHegde commented:
@damithc Ready for review.

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackAddAction.java:
```diff
@@ -147,88 +143,19 @@ private void createTemplateFeedbackQuestions(String courseId, String feedbackSes
         newSession.setSentOpenEmail(false);
         newSession.setSentPublishedEmail(false);
-        
         newSession.setFeedbackSessionType(FeedbackSessionType.STANDARD);
```
This is common to both this action and the other action. Can move to the abstract action instead?

AnaghHegde commented:
@wkurniawan07 
>In what way does having just one method affect code readability?
I didn't give a thought about two methods but i was talking about `extractFeedbackSessionData` where the attributes order will change( earlier one order now in different order). 

> failing tests is not a good reason. "It matters": in what ways do it matter?
as of now i m not able to figure out why exactly its failing.
but in build log i don't understand this one 
```
<expected>feedbackSessionType=PRIVATE where as in the method its set to feedbackSessionType=STANDARD 
```

>I don't get what this means.
See line 113 of new `*Abstraction` class. This `List` contains `ReminderEmailsList` - list people to whom the remainder email regarding feedback has to be sent. If this `list` is created at the beginning of the method `list` will have contains people - who already submitted feedback (line 114 which checks for null). Isn't it ?

>Which lines give you this impression?
Seeing the build failure log i m thinking of that.

About the failing test cases is there any wrong in the way i have added all to one method ?.

AnaghHegde commented:
@wkurniawan07 Can you please help me ? or little guide in what way should i have to proceed ?

AnaghHegde commented:
@wkurniawan07 Sorry to say this but its been 9 days.

wkurniawan07 commented:
> About the failing test cases is there any wrong in the way i have added all to one method ?

It might be. It might be the case that it's not possible at all, or it might just be that you have not done it right. The problem is, your reasonings are not solid enough. "Test failure" or "Build failure" are not valid reason for things to not be done a certain way (or else many parts of this application will not be as optimal as it is right now). You need to learn not to give up that easily even though the CI may not be in your favour.

AnaghHegde commented:
@wkurniawan07 Ready for review. Please suggest a better name for the helper functions.

AnaghHegde commented:
@wkurniawan07 Ready for review.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditSaveAction.java:
```diff
@@ -64,93 +60,14 @@ protected ActionResult execute() throws EntityDoesNotExistException {
-        newSession.setPublishedEmailEnabled(sendReminderEmailsList.contains(EmailType.FEEDBACK_PUBLISHED.toString()));
-
+        newSession.setInstructions(new Text(
+                getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_INSTRUCTIONS)));
```
Make these on the same line if possible

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditSaveAction.java:
```diff
@@ -64,93 +60,14 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+        newSession.setInstructions(new Text(
+                getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_INSTRUCTIONS)));
+        newSession.setOpeningEmailEnabled(sendReminderEmailsList.contains(EmailType.FEEDBACK_OPENING
+                .toString()));
```
Here as well.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAddAction.java:
```diff
@@ -146,87 +141,13 @@ private void createTemplateFeedbackQuestions(String courseId, String feedbackSes
-        newSession.setPublishedEmailEnabled(sendReminderEmailsList.contains(EmailType.FEEDBACK_PUBLISHED.toString()));
-
+        newSession.setInstructions(new Text(
+                getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_INSTRUCTIONS)));
```
Make these on the same line if possible

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,155 @@
+            List<InstructorAttributes> instructorList) {
+
+        List<FeedbackSessionAttributes> sessions =
+                logic.getFeedbackSessionsListForInstructor(instructorList);
```
Make these on the same line if possible

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    protected List<FeedbackSessionAttributes> loadFeedbackSessionsList(
+            List<InstructorAttributes> instructorList) {
```
Make these on the same line if possible

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,155 @@
+        return this.extractFeedbackSessionDataHelper(newSession, sendRemainderEmailsList);
+    }
+
+    protected FeedbackSessionAttributes extractFeedbackSessionData() {
```
You may want to refactor (break up) this method into smaller methods

AnaghHegde commented:
@samsontmr @wkurniawan07 Ready for review. 

@samsontmr I have refactored method . Please check.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,164 @@
+        newSession.setFeedbackSessionName(SanitizationHelper.sanitizeTitle(
+                getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME)));
+        newSession.setFeedbackSessionType(FeedbackSessionType.STANDARD);
+        // newSession = extractFeedbackSessionDataHelper(newSession);
```
Remove this line if it's no longer needed.

samsontmr replied:
This method looks like it can still be broken up further.

AnaghHegde replied:
Yeah it can be but its not of similar types(like the previous one). But ok i will refactor it.

Neurrone commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,164 @@
+    protected FeedbackSessionAttributes extractFeedbackSessionDataHelper(
+            FeedbackSessionAttributes newSession, List<String> sendRemainderEmailsList) {
+        return this.extractFeedbackSessionDataHelper(newSession, sendRemainderEmailsList);
+    }
```
What's the purpose of this method?

AnaghHegde replied:
Instead of classname for comparison i am using inheritance concept.

wkurniawan07 replied:
This is a dangerous approach. If you call this method without overriding it, you're run into an infinite loop.
What you want to do is perhaps `abstract` method.

Neurrone replied:
Did you mean to use polymorphism here to define some abstract method called extractFeedbackSessionDataHelper, and have it implemented by child classes? extractFeedbackSessionDataHelper is not a very good name, as it doesn't give any indication of what its doing.

AnaghHegde replied:
>extractFeedbackSessionDataHelper 

This is temporary i am working on a better name.

>What you want to do is perhaps abstract method

@wkurniawan07 The problem is i can't have any definition in an abstract method right?

wkurniawan07 replied:
> The problem is i can't have any definition in an abstract method right?

Why would you need a definition in the first place?

AnaghHegde replied:
Yeah actually that works but if i make it abstract i have inherit it in `InstructorFeedbackAbstractAction` also as it also extends the `InstructorFeedbackAbstractAction` .

Neurrone replied:
Making it abstract isn't what you want then. What're you trying to do in those helper methods? You can move the common functionality up to the AbstractAction so that those child classes that need it can use it. Right now, that definition of the helper method in the AbstractAction isn't useful, or dangerous at worst.

AnaghHegde replied:
>What're you trying to do in those helper methods? 

The non common functionalities are there in that method.

>You can move the common functionality up to the AbstractAction so that those child classes that need it can use it

That is what i have done. 

> Right now, that definition of the helper method in the AbstractAction isn't useful, or dangerous at worst.

Earlier i used class names for comparison. But most said to use polymorphism as class names can be changed in the future. 

So what are you suggesting.

wkurniawan07 replied:
> i have inherit it in InstructorFeedbackspageAction also as it also extends the InstructorFeedbackAbstractAction

Then make the method unusable in that class.
```java
@Override
protected FeedbackSessionAttributes extractFeedbackSessionDataHelper(FeedbackSessionAttributes newSession, List<String> sendRemainderEmailsList) {
    throw new UnsupportedOperationException("Method is not used here");
    // or simply:
    // return null; // method is not used here
}
```
and do rename the method. `Helper` does not tell anything about what it does.

AnaghHegde replied:
this also works. I am thinking of making the class `abstract` but i don't think thats a good idea.

AnaghHegde commented:
@samsontmr Ready for review. I think i might have done too much refactoring.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,170 @@
+        newSession.setFeedbackSessionName(SanitizationHelper.sanitizeTitle(
+                getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME)));
+        newSession.setFeedbackSessionType(FeedbackSessionType.STANDARD);
+        newSession = setTime(newSession);
```
I don't think you need to return `newSession` each time you call a `set*` method.

AnaghHegde commented:
@samsontmr Ready for review.

@wkurniawan07 I made the method as abstract. I changed the method name also(But i am not sure if it gives expected meaning or not).

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbacksPageAction.java:
```diff
@@ -66,4 +66,9 @@ protected ActionResult execute() {
+    @Override
+    protected FeedbackSessionAttributes extractFeedbackSessionDataSetUniqueAttributes(
+            FeedbackSessionAttributes newSession, List<String> sendRemainderEmailsList) {
+        throw new UnsupportedOperationException("Method is not used here");
```
Perhaps "Method should not be used here" would sound more appropriate?

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAddAction.java:
```diff
@@ -141,7 +141,8 @@ private void createTemplateFeedbackQuestions(String courseId, String feedbackSes
 
-    protected FeedbackSessionAttributes extractFeedbackSessionDataHelper(
+    @Override
+    protected FeedbackSessionAttributes extractFeedbackSessionDataSetUniqueAttributes(
```
Is the `extractFeedbackSessionData` part of the method name necessary? It seems like you're simply setting attributes.

AnaghHegde replied:
Yeah but if i just give  `SetUniqueAttributes` wouldn't it be difficult in sub classes to make out what's happening ?.

samsontmr replied:
`setUniqueAttributesForSession`? Also, correct me if I'm wrong, but I don't think you need a return. And the argument should be `session` and not `newSession`.

AnaghHegde replied:
>setUniqueAttributesForSession? Also, correct me if I'm wrong, but I don't think you need a return.

My bad didn't remove it.

AnaghHegde commented:
@samsontmr @wkurniawan07 Ready for review. I have fixed the name and returns type.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    private void setTime(FeedbackSessionAttributes newSession) {
+
```
Remove this empty line.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,165 @@
+            } catch (NumberFormatException nfe) {
+                log.warning("Failed to parse time zone parameter: " + paramTimeZone);
+            }
+        }
```
This block could be extracted into its own method.

AnaghHegde replied:
ok you mean just the try catch block ?

samsontmr replied:
If `paramTimeZone` is not used elsewhere, from line 36 onward.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,165 @@
+            newSession.setGracePeriod(Integer.parseInt(paramGracePeriod));
+        } catch (NumberFormatException nfe) {
+            log.warning("Failed to parse graced period parameter: " + paramGracePeriod);
+        }
```
Same for this.

AnaghHegde replied:
same as above

samsontmr replied:
Same here.

samsontmr replied:
Hmm sorry if I wasn't clear enough, `setTimeZone` and `setGracePeriod` should be two separate methods.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    protected List<CourseAttributes> loadCoursesList(List<InstructorAttributes> instructorList) {
+
```
Remove empty line.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    protected List<FeedbackSessionAttributes> loadFeedbackSessionsList(List<InstructorAttributes> instructorList) {
+
```
Remove empty line.

samsontmr commented:
@AnaghHegde Try to have more meaningful commit messages. This helps the reviewers. You can also consider making one commit per logical change, I find that it helps when things go wrong and you need to roll back. This helps the reviewers as well.

AnaghHegde commented:
@samsontmr Ok. I will not repeat this in future.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,161 @@
+        // handle session visible after results visible to avoid having a
+        // results visible date when session is private (session not visible)
+        type = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_SESSIONVISIBLEBUTTON);
+        setSessionVisibleFromTime(newSession, type);
```
Same here.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,161 @@
+        setTime(newSession);
+        setTimeZoneGracePeriod(newSession);
+        String type = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_RESULTSVISIBLEBUTTON);
+        setResultsVisibleFromTime(newSession, type);
```
Line 36 can be inside this method instead of passing `type` in as an argument.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,161 @@
+        } catch (NumberFormatException nfe) {
+            log.warning("Failed to parse graced period parameter: " + paramGracePeriod);
+        }
+
```
Unnecessary space here can be removed.

samsontmr replied:
Unnecessary empty line.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,161 @@
+
+    private void setResultsVisibleFromTime(FeedbackSessionAttributes newSession) {
+        String type = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_RESULTSVISIBLEBUTTON);
+        switch (type) {
```
Since type is only used once, I recommend inlining the method call, e.g. 
`switch (getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_RESULTSVISIBLEBUTTON))`. Same goes for the other similar method.

AnaghHegde replied:
don't you think its costlier than the previous operation? as we have to call the method every time.

wkurniawan07 replied:
@AnaghHegde, that is not how caching works.

AnaghHegde commented:
@samsontmr Ready for review.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,158 @@
+        }
+        return courseInstructorMap;
+    }
+
```
Unnecessary empty line.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,158 @@
+        return newSession;
+    }
+
+    private void setTimeZoneGracePeriod(FeedbackSessionAttributes newSession) {
```
`setTimeZone` and `setGracePeriod` should be two separate methods.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,158 @@
+            break;
+        default:
+            log.severe("Invalid resultsVisibleFrom setting in creating"
+                       + newSession.getIdentificationString());
```
Combine into single line if possible.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,158 @@
+            break;
+        default:
+            log.severe("Invalid sessionVisibleFrom setting in creating "
+                       + newSession.getIdentificationString());
```
Same here.

samsontmr commented on src/main/java/teammates/ui/controller/InstructorFeedbackAbstractAction.java:
```diff
@@ -0,0 +1,158 @@
+    protected Map<String, InstructorAttributes> loadCourseInstructorMap(boolean omitArchived) {
+        Map<String, InstructorAttributes> courseInstructorMap = new HashMap<String, InstructorAttributes>();
+        List<InstructorAttributes> instructors = logic.getInstructorsForGoogleId(account.googleId,
+                                                                                 omitArchived);
```
Change this to
```
List<InstructorAttributes> instructors = 
        logic.getInstructorsForGoogleId(account.googleId, omitArchived);
```

samsontmr commented:
This PR is getting a little long, perhaps you can reorganise your commits the way I suggested in your other ongoing PR?

AnaghHegde commented:
@samsontmr Shall i open a new PR as this PR is too long?.

samsontmr commented:
> Shall i open a new PR as this PR is too long?.

Not necessary if you reorganize your commits the way I suggested.

AnaghHegde commented:
From which commit do i have to reorganise the commits?

samsontmr commented:
> From which commit do i have to reorganise the commits?

From the first commit.

AnaghHegde commented:
>From the first commit.

It has total 107 commits as of now.

samsontmr commented:
> It has total 107 commits as of now.

That is precisely why it needs to be reorganized.

damithc commented:
@AnaghHegde you can use [this](https://github.com/oss-generic/process/blob/master/docs/FormatsAndConventions.md#commit-organization) for guidance when organizing your commits.

AnaghHegde commented:
@samsontmr ok just give me some time. I will work on it.

AnaghHegde commented:
@samsontmr This PR is getting very long. So i will open a new PR with commits organised from the 1st commit. Sorry for the inconvenience.
